### PR TITLE
Add Preference Graph

### PR DESCRIPTION
**Pull request by moririn2528**
## Contributor License Agreement

This repository (``optuna-dashboard``) and [Goptuna](https://github.com/c-bata/goptuna) share common code.
This pull request may therefore be ported to Goptuna.
Make sure that you understand the consequences concerning licenses and check the box below if you accept the term before creating this pull request.

* [x] I agree this patch may be ported to [Goptuna](https://github.com/c-bata/goptuna) by other Goptuna contributors.

## What does this implement/fix? Explain your changes.
I added graphs of preferences.
スクリーンショット 2023-09-08 15 37 50
スクリーンショット 2023-09-08 15 37 58

### CODE REVIEW

File: webpack.config.js [hunk 1]
```diff
@@ -1,65 +1,81 @@
...
+    filename: "bundle.js",
+    publicPath: "/public/",
+  },
+  module: {
+    rules: [
+      { oneOf: [typeScriptLoader] },
+      {
+        test: /\.css$/,
+        use: ["style-loader", "css-loader"],
+      },
```
Comment (moririn2528) [inline webpack.config.js:43]: In this file, I change only this block. Others are changed by formatter.

File: optuna_dashboard/ts/components/AppDrawer.tsx [hunk 2]
```diff
@@ -246,6 +248,24 @@ export const AppDrawer: FC<{
...
+                  component={Link}
+                  to={`${URL_PREFIX}/studies/${studyId}/graph`}
+                  sx={styleListItemButton}
+                  selected={page === "graph"}
+                >
+                  <ListItemIcon sx={styleListItemIcon}>
+                    <LanIcon />
+                  </ListItemIcon>
+                  <ListItemText
+                    primary="PreferenceGraph"
```
Comment (c-bata) [inline optuna_dashboard/ts/components/AppDrawer.tsx:263]: How about this?

```suggestion
                    primary="Preferences (Graph)"
```

Comment (c-bata) [reply optuna_dashboard/ts/components/AppDrawer.tsx:263]: I actually do not prefer "PreferenceHistory" and "HumanInTheLoop". At least there should be a space between each word. What do you think?

File: optuna_dashboard/ts/types/index.d.ts [hunk 3]
```diff
@@ -223,3 +223,5 @@ type PreferenceHistory = {
   feedback_mode: PreferenceFeedbackMode
   timestamp: Date
 }
+
+declare module "*.css"
```
Comment (c-bata) [inline optuna_dashboard/ts/types/index.d.ts:227]: Could you move this to the top of this file?

File: optuna_dashboard/ts/components/PreferentialGraph.tsx [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+        )
+      })
+      .catch(console.error)
+    setEdges(
+      preferences.map((p) => {
+        return {
+          id: `e${p[0]}-${p[1]}`,
+          source: `${p[0]}`,
+          target: `${p[1]}`,
+          style: { stroke: isDarkMode ? "#fff" : "#000" },
```
Comment (c-bata) [inline optuna_dashboard/ts/components/PreferentialGraph.tsx:247]: Can we use theme.palette.background color instead?

**Review by c-bata — CHANGES_REQUESTED** [review]
Thank you for your pull request. I left some comments.

File: optuna_dashboard/ts/components/PreferentialGraph.tsx [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+const nodeTypes: NodeTypes = {
+  note: GraphNode,
+}
+const defaultEdgeOptions: DefaultEdgeOptions = {
+  animated: true,
+}
+
+function reductionPreference(
+  input_preferences: [number, number][]
+): [number, number][] {
```
Comment (c-bata) [inline optuna_dashboard/ts/components/PreferentialGraph.tsx:99]: How about using arrow function expression?

File: optuna_dashboard/ts/components/AppDrawer.tsx [hunk 6]
```diff
@@ -246,6 +248,24 @@ export const AppDrawer: FC<{
                 <ListItemText primary="Analytics" sx={styleListItemText} />
               </ListItemButton>
             </ListItem>
+            {studyDetail?.is_preferential && (
```
Comment (c-bata) [inline optuna_dashboard/ts/components/AppDrawer.tsx:251]: How about checking `studySummary?.is_preferential` as well? Perhaps it is better to define the function like below.

https://github.com/optuna/optuna-dashboard/blob/238dffa80f347081fb93b77436c9c21dd4153439/optuna_dashboard/ts/state.ts#L82-L88

**Review by c-bata — COMMENTED** [review]
Looks almost good to me. Left some comments.

**Review by c-bata — APPROVED** [review]
LGTM!

**Pull request merged**